aezeed/cipherseed.go

```diff
@@ -0,0 +1,484 @@
+	// KDF to obtain the key we'll use for encryption.
+	key, err := scrypt.Key(
+		pass, c.salt[:], scryptN, scryptR, scryptP, keyLen,
+	)
+	if err != nil {
+		return cipherSeedBytes, err
+	}
+
+	// Next, we'll encode the serialized plaintext cipherseed into a buffer
+	// that we'll use for encryption.
+	seedBytes := bytes.NewBuffer(make([]byte, 0, DecipheredCipherSeedSize))
+	if err := c.encode(seedBytes); err != nil {
+		return cipherSeedBytes, err
+	}
+
+	// With our plaintext seed encoded, we'll now construct the AD that
+	// will be passed to the encryption operation. This ensures to
+	// authenticate both the salt and the external version.
+	ad := encodeAD(CipherSeedVersion, c.salt)
+
+	// With all items assembled, we'll now encipher the plaintext seed
+	// with our AD, key, and MAC size.
+	cipherSeed := seedBytes.Bytes()
+	cipherText := aez.Encrypt(
+		key, nil, [][]byte{ad[:]}, CipherTextExpasnion, cipherSeed, nil,
+	)
+
+	// Finally, we'll pack the {version || ciphertext || salt| seed into a
+	// byte slice for encoding as a mnemonic.
+	cipherSeedBytes[0] = byte(CipherSeedVersion)
+	copy(cipherSeedBytes[1:28], cipherText)
+	copy(cipherSeedBytes[28:], c.salt[:])
+
+	return cipherSeedBytes, nil
+}
+
+// cipherTextToMnemonic converts the aez ciphertext appended with the salt to a
+// 24-word mnemonic pass phrase.
+func cipherTextToMnemonic(cipherText [EncipheredCipherSeedSize]byte) (Mnenonic, error) {
+	var words [NumMnenonicWords]string
+
+	// First, we'll convert the ciphertext itself into a bitstream for easy
+	// manipulation.
+	cipherBits := bstream.NewBStreamReader(cipherText[:])
+
+	// With our bitstream obtained, we'll read 11 bits at a time, then use
+	// that to index into our word list to obtain the next word.
+	for i := 0; i < NumMnenonicWords; i++ {
+		index, err := cipherBits.ReadBits(bitsPerWord)
+		if err != nil {
+			return words, nil
+		}
+
+		words[i] = defaultWordList[index]
+	}
+
+	return words, nil
+}
+
+// ToMnemonic maps the final enciphered cipher seed to a human readable 24-word
+// mnenonic phrase. The password is optional, as if it isn't specified aezeed
+// will be used in its place.
+func (c *CipherSeed) ToMnemonic(pass []byte) (Mnenonic, error) {
+	// If the passphrase wasn't provided, then we'll use the string
+	// "aezeed" in place.
+	passphrase := pass
+	if len(passphrase) == 0 {
+		passphrase = []byte("aezeed")
+	}
+
+	// First, we'll convert the valid seed triple into an aez cipher text
+	// with our KDF salt appended to it.
+	cipherText, err := c.encipher(pass)
+	if err != nil {
+		return Mnenonic{}, nil
+	}
+	c.cipherText = cipherText
+
+	// Now that we have our cipher text, we'll convert it into a mnenonic
+	// phrase.
+	return cipherTextToMnemonic(cipherText)
+}
+
+// Encipher maps the cipher seed to an aez ciphertext using an optional
+// passphrase.
+func (c *CipherSeed) Encipher(pass []byte) ([EncipheredCipherSeedSize]byte, error) {
+	return c.encipher(pass)
```

Comment:
Shouldn't we cache the ciphertext here as well?

Reply:
Also, we'll want to check if a password was provided like in `ToMnemonic`. It might be better to move the check into `encipher` to avoid having to do it in both `Encipher` and `ToMnemonic`.

Reply:
Fixed.